### ISSUE TITLE
Anchor Events client error when handling transaction with multiple instructions

### ISSUE THREAD

**Issue by tsfotis** [otter-sec/anchor#1941]
The anchor client when subscribed to program logs fails with error

```bash
thread '<unnamed>' panicked at 'assertion failed: !self.stack.is_empty()', /anchor-client-0.24.2/src/lib.rs:359:9
```

Steps to reproduce

-   init anchor PubsubClient client
-   send a successful transaction with multiple instructions, at least 1 from the program

It seems like the [Execution](https://github.com/project-serum/anchor/blob/master/client/src/lib.rs#L334) is able to handle logs only from a transaction with one instruction.

**Comment by yourarj** [comment]
I'll have a look and let you know if any further clarification is needed. Thanks

**Comment by yourarj** [comment]
@tsfotis when an event is dispatched the stack (vec) in Execution is popped. No matter how many instruction currently vec is programmed to have only one entry technically to dispatch only one event. 

Solution approach:
Add program id (str) entry per instruction in Execution stack. (I think it makes more sense)

Is this the expected solution?

**Comment by yourarj** [comment]
@tsfotis please check the draft PR and have quick discussion.

**Comment by mgild** [comment]
This still seems to be an error surfacing on anchor 26

**Comment by Henry-E** [comment]
good point, the PR had to be reverted. If anyone would like to make a fix that passes test happy to check it out.

https://github.com/coral-xyz/anchor/pull/2282

**Comment by mgild** [comment]
Getting things started here for how I handle it now:
```
    loop {
        match r.recv() {
            Ok(y) => {
                let log: String = y.value.logs.join(" ");
                let words: Vec<&str> = log.split(" ").filter(|s| s.len() >= 8).collect();
                for w in words {
                    let decoded = general_purpose::STANDARD.decode(w);
                    if decoded.is_err() {
                        continue;
                    }
                    let decoded = decoded.unwrap();
                    if decoded[..8] != AggregatorOpenRoundEvent::DISCRIMINATOR {
                        continue;
                    }
                    let event = AggregatorOpenRoundEvent::try_from_slice(&decoded[8..]);
                    if event.is_ok() {
                        println!("{:?}", event.unwrap());
                    }
                }
            }
            Err(_) => {
                println!("Channel closed");
                break;
            }
        }
    }
```

**Comment by mgild** [comment]
Might need to ensure that putting the pid in "Mentions" really only targets logs of the intended program

**Comment by mgild** [comment]
This should probably parse only program log lines in the scope of the intended program invoke too

**Comment by yourarj** [comment]
@mgild I'll try to take it up this week if time permits

**Comment by yourarj** [comment]
@mgild @tsfotis @Henry-E have raised another Draft PR. https://github.com/coral-xyz/anchor/pull/2554

Would like to have your opinion on the approach.

Especially https://github.com/coral-xyz/anchor/pull/2554/files#diff-3f7e738903cc8204a4e3b1c8ab4db5812898e6bb5ad8b6c7b36df42e29481a8fR356

We try to get program id earlier before it is being pushed by handle_system_log. Hence preprocessed it in execution construction and ignore them in handle_system_log

**Comment by yourarj** [comment]
> > @tsfotis can you add more details and code where you can reproduce this? That'll be helpful to jumpstart.
> 
> @yourarj [here](https://github.com/tsfotis/anchor-sub-client-issue) is an example. In the test you can see that when multiple instructions are send in a single transaction we get the above error.

Have also integrated with this example and verified
